.docker/Dockerfile

```diff
@@ -0,0 +1,291 @@
+# syntax=docker/dockerfile:1.7
+ARG DEV_FROM_STAGE=tooler
+ARG EXPORT_FROM_STAGE=builder
+ARG PREP_FROM_STAGE=runner
+ARG SHIP_FROM_STAGE=runner
+ARG WS_CACHE_ID=nav2
+
+ARG FROM_IMAGE=base
+# Stage from full image tag name for dependabot detection
+FROM ros:rolling AS base
+
+################################################################################
+# MARK: baser - setup base image using snapshots
+################################################################################
+FROM $FROM_IMAGE AS baser
+ENV FROM_IMAGE=${FROM_IMAGE}
+
+# Configure ubuntu snapshot
+RUN UBUNTU_DEB_SNAPSHOT=$(date -r /var/lib/dpkg/info +%Y%m%dT%H%M%SZ) && \
+  sed -i "s|http://archive.ubuntu.com/ubuntu/|http://snapshot.ubuntu.com/ubuntu/${UBUNTU_DEB_SNAPSHOT}|g" /etc/apt/sources.list && \
+  sed -i "s|http://security.ubuntu.com/ubuntu/|http://snapshot.ubuntu.com/ubuntu/${UBUNTU_DEB_SNAPSHOT}|g" /etc/apt/sources.list
+
+# Edit apt config for caching
+RUN mv /etc/apt/apt.conf.d/docker-clean /etc/apt/ && \
+    echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' \
+      > /etc/apt/apt.conf.d/keep-cache && \
+    # Given fixed snapshots, just cache apt update once
+    apt-get update && echo v1
+
+# Configure overlay workspace
+ENV OVERLAY_WS=/opt/nav2_ws
+WORKDIR $OVERLAY_WS
+
+# install bootstrap tools
+RUN --mount=type=cache,sharing=locked,target=/var/cache/apt \
+    apt-get install -y --no-install-recommends \
+      gettext-base \
+      wget \
+      zstd
+
+# init and update rosdep
+ENV ROS_HOME=/opt/.ros
+RUN rosdep update \
+      --rosdistro $ROS_DISTRO
+
+################################################################################
+# MARK: cacher - cache source and dependency instructions
+################################################################################
+FROM baser AS cacher
+
+# copy overlay source
+COPY ./ ./src/navigation2
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+RUN dep_types=(\
+      "exec:--dependency-types=exec" \
+      "test:--dependency-types=test" \
+      "build:"\
+    ) && \
+    for dep_type in "${dep_types[@]}"; do \
+      IFS=":"; set -- $dep_type; \
+      rosdep install -y \
+        --from-paths src \
+        --ignore-src \
+        --skip-keys " \
```

Comment:
Worth having a skipped keys txt as well?

Reply:
Indeed, I'll try and consolidate this again.